### ISSUE TITLE
Streams.forEachWithIndex(*Stream, *ConsumerWithIndex)?

### ISSUE THREAD

**Issue by perceptron8** [google/guava#2913]
Does the following statement from #35 also apply to Streams?

> The "enumerate" idea is one we considered and rejected long ago;

I'd like to ask for / propose some kind of intersection between `Streams.forEachPair` and `Streams.mapWithIndex`, let's name it `Streams.forEachWithIndex`.

Here's a quick and dirty prototype:

```
@Beta
public interface ConsumerWithIndex<T> {
  /** Performs this operation on the given argument and its index within a stream. */
  void accept(T value, long index);
}

@Beta
public static <T> void forEachWithIndex(Stream<T> stream, ConsumerWithIndex<T> consumer) {
  checkNotNull(consumer);
  if (stream.isParallel()) {
    // TODO eliminate (un)boxing and TemporaryPair creation?
    forEachPair(stream, LongStream.rangeClosed(0, Long.MAX_VALUE).boxed(),
        (T value, Long index) -> consumer.accept(value, index));
  } else {
    long index = 0;
    Iterator<T> iter = stream.iterator();
    while (iter.hasNext()) {
      consumer.accept(iter.next(), index++);
    }
  }
}
```

I admit that this method can be used incorrectly (imagine looping-with-index over `Set.stream`), but the same concern applies to `mapWithIndex` and `forEachPair`.

If you have collection, you can just use for loop. In case of streams (esp. with no backing collection), size may be unknown - and you have declare counter, iterate and increment - and that's a lot of ceremony. Also, counter must have a wider scope than your loop...

This may be not the most needed method in the world, but I missed it several times since Guava 21 was released.

What do you think?

**Comment by orionll** [comment]
`ConsumerWithIndex` is unnecessary. There is already `ObjLongConsumer<T>` in java.util.function.

**Comment by perceptron8** [comment]
@orionll True, but I was trying to follow reasoning from https://github.com/google/guava/wiki/IdeaGraveyard#functionspredicates-for-n--2-inputs. `Streams.ConsumerWithIndex` would be more specialized than `ObjLongConsumer`, and it's name would be analogous to `Streams.FunctionWithIndex`.

**Comment by orionll** [comment]
@perceptron8 FunctionWithIndex exists only because there is no ObjLongBiFunction in Java. However, ObjLongConsumer does exist, so I see no good reason to create a duplicate.

**Comment by benmccann** [comment]
@lowasser would you be open to me sending a PR implementing this functionality?
